### PR TITLE
Fix documentation metadata getting lost

### LINKED ISSUE #434: Documentation Metadata is Lost at the End of the Compilation
### General Summary

`DocumentationComments` compiler pass attaches documentation to the node metadata. When the node is rewritten, metadata is lost.

### Steps to Reproduce
``` scala
implicit val passManager: PassManager = 
  new Passes().passManager
implicit val moduleContext: ModuleContext = 
  ModuleContext(freshNameSupply = Some(new FreshNameSupply))

val module =
  """## The foo
    |foo = 42""".stripMargin.preprocessModule
assert(module.bindings(0).getMetadata(DocumentationComments).isDefined)
```

### Expected Result

Documentation is preserved after all compiler passes.

### Actual Result

Documentation is lost.

### Enso Version

Commit: main @  f7d4ef546aa43a3cb084264bc69a8b9d7643ef95

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Fixes #434 

### Important Notes

Additionally, adds support for documentation comments on pattern match branches and diagnostic deduplication in the `GatherDiagnostics` pass.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/pass/desugar/NestedPatternMatch.scala [hunk 1]
```diff
@@ -346,33 +349,40 @@ case object NestedPatternMatch extends IRPass {
...
-      }
-  }
+  def containsNestedPatterns(pattern: Pattern): Boolean =
+    pattern match {
+      case _: Pattern.Name => false
+      case Pattern.Constructor(_, fields, _, _, _) =>
+        fields.exists {
+          case _: Pattern.Constructor => true
+          case _: Pattern.Name        => false
+          case _: Pattern.Doc         => false
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/desugar/NestedPatternMatch.scala]: This occurring here should be a fatal error. Same for the other cases below.

File: engine/runtime/src/main/scala/org/enso/compiler/pass/optimise/ApplicationSaturation.scala [hunk 2]
```diff
@@ -192,37 +192,30 @@ case object ApplicationSaturation extends IRPass {
   type KnownFunctionsMapping = Map[String, FunctionSpec]
 
   /** Describes the saturation state of a function application. */
-  sealed trait CallSaturation extends IRPass.Metadata
+  sealed trait CallSaturation extends IRPass.Metadata {
+    // TODO [RW] CallSaturation is immutable, right?
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/optimise/ApplicationSaturation.scala]: Correct!

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala [hunk 3]
```diff
@@ -4484,6 +4485,87 @@ object IR {
...
+           |id = $id
+           |)
+           |""".toSingleLine
+
+      /** Shows the IR as code.
+        *
+        * @param indent the current indentation level
+        * @return a string representation of `this`
+        */
+      override def showCode(indent: Int): String = s"## $doc"
```
[the remark this answers is not in the document]
Comment (radeusgd) [reply engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala:4590]: Seems like it is preserved (although when I checked it, it seems there is one space less than in source file (weird ?)). The `showCode` for `Comment.Documentation` is implemented in exactly the same way though.

File: build.sbt [hunk 4]
```diff
@@ -575,10 +579,10 @@ lazy val searcher = project
   .configs(Test)
   .settings(
     libraryDependencies ++= Seq(
-      "com.typesafe.slick" %% "slick"       % "3.3.2",
-      "org.xerial"          % "sqlite-jdbc" % "3.31.1",
-      "org.scalatest"      %% "scalatest"   % scalatestVersion % Test,
-    )
+        "com.typesafe.slick" %% "slick"       % "3.3.2",
```
Comment (iamrecursion) [inline build.sbt]: Please hoist these versions to the top as with all others.

**Review by iamrecursion — APPROVED** [review]

**Pull request merged**